epsie/proposals/birthdeath.py

```diff
@@ -0,0 +1,315 @@
+        # store parameters
+        self.update_parameters = update_parameters
+        self.model_parameter = bd_proposal.parameter
+        self.parameters = update_parameters + [self.model_parameter]
+        # store proposals
+        self.update_proposals = list(update_proposals)
+        self.bd_proposal = bd_proposal
+        self.all_proposals = self.update_proposals + [self.bd_proposal]
+        # the proposal is symmetric only if all of the constitutent
+        # proposals are also
+        self._symmetric = all([prop.symmetric
+                              for prop in self.all_proposals])
+        # store the prior distribution to sample new components on the go
+        self._prior_dist = None
+        self.prior_dist = prior_dist
+        # set the bit generator
+        self.bit_generator = bit_generator
+        # have all of the proposals use the same random state
+        for prop in self.all_proposals:
+            prop.bit_generator = self.bit_generator
+
+        self._last_prop = None
+
+
+    @property
+    def symmetric(self):
+        return self._symmetric
+
+    def logpdf(self, xi, givenx):
+        lp = 0.0
+        for prop in self.all_proposals:
+            if prop.active:
+                lp += prop.logpdf({p: xi[p] for p in prop.parameters},
+                                  {p: givenx[p] for p in prop.parameters})
+        return lp
+
+    def update(self, chain):
+        # update each of the proposals
+        acc = chain.acceptance[-1]['accepted']
+        if acc and (self._last_prop is not None):
+            self._last_prop.active = not self._last_prop.active
+
+        self._last_prop = None
+
+        # turn this back on later but need to check what happens to adaptation
+        # if proposals are getting turned off and on. Do adaptation only on
+        # proposals that just did a within-model MCMC step?
+        # Also what we could do is introduce an adaption of the mean of the
+        # poisson distribution that decides where to jump next.
+
+#        for prop in self.all_proposals:
+#            prop.update(chain)
+
+    @property
+    def prior_dist(self):
+        return self._prior_dist
+
+    @prior_dist.setter
+    def prior_dist(self, prior_dist):
+        try:
+            self._prior_dist = {p: prior_dist[p]
+                                for p in self.update_parameters}
+        except KeyError:
+            raise ValueError('provide a value for each parameter')
+
+    def jump(self, fromx):
+        givenk = fromx[self.model_parameter]
+        newk = self.bd_proposal.jump({self.model_parameter: givenk})
+        # unpack it
+        newk = newk[self.model_parameter]
+
+        inactive_props = list()
+        active_props = list()
+        for prop in self.update_proposals:
+            if prop.active:
+                active_props.append(prop)
+            else:
+                inactive_props.append(prop)
```

Comment:
This will be faster as a list comprehension...
```
inactive_props = [prop for prop in self.update_proposals if not prop.active]
active_props = [prop for prop in self.update_proposals if prop.active]
```